helix-vcs/src/git/blame.rs

```diff
@@ -0,0 +1,628 @@
+        // - cursor_line is 8
+        //
+        // So when our cursor is on the 10th added line or earlier, blame_line will be 0. This means
+        // the blame will be incorrect. But that's fine, because when the cursor_line is on some hunk,
+        // we can show to the user nothing at all. This is detected in the editor
+        let blame_line = line.saturating_sub(inserted_lines) + removed_lines;
+        let repo = self.repo.to_thread_local();
+
+        let commit = self
+            .blame
+            .get(&blame_line)
+            .and_then(|obj| repo.find_commit(*obj).ok());
+        let message = commit.as_ref().and_then(|c| c.message().ok());
+        let author = commit.as_ref().and_then(|c| c.author().ok());
+
+        LineBlame {
+            commit_hash: commit
+                .as_ref()
+                .and_then(|c| c.short_id().map(|id| id.to_string()).ok()),
+            author_name: author.map(|a| a.name.to_string()),
+            author_email: author.map(|a| a.email.to_string()),
+            commit_date: author.map(|a| a.time.format(gix::date::time::format::SHORT)),
+            commit_message: message.as_ref().map(|msg| msg.title.to_string()),
+            commit_body: message
+                .as_ref()
+                .and_then(|msg| msg.body.map(|body| body.to_string())),
+            time_ago: author
+                .map(|a| helix_stdx::date::format_relative_time(a.time.seconds, a.time.offset)),
+        }
+    }
+
+    /// Compute blame for this file
+    pub fn try_new(file: PathBuf) -> Result<Self> {
+        let thread_safe_repo =
+            open_repo(get_repo_dir(&file)?).context("failed to open git repo")?;
+        let repo = thread_safe_repo.to_thread_local();
+        let head = repo.head()?.peel_to_commit_in_place()?.id;
+
+        let traverse = gix::traverse::commit::topo::Builder::from_iters(
+            &repo.objects,
+            [head],
+            None::<Vec<gix::ObjectId>>,
+        )
+        .build()?;
+
+        let relative_path = file
+            .strip_prefix(
+                thread_safe_repo
+                    .path()
+                    .parent()
+                    .context("Could not get parent path of repository")?,
+            )
+            .unwrap_or(&file)
+            .to_str()
+            .context("Could not convert path to string")?;
+
+        let mut resource_cache = repo.diff_resource_cache_for_tree_diff()?;
+        let file_blame = gix::blame::file(
+            &repo.objects,
+            traverse.into_iter(),
+            &mut resource_cache,
+            // FIXME: gix v0.70 does not account for backslashes in Windows and hard-codes
+            // the `/` as the path separator.
+            // There is a PR to fix that (https://github.com/GitoxideLabs/gitoxide/pull/1905#issue-2939833001)
+            // so we should remove this conversion once we update to a newer Gix version
+            BStr::new(&relative_path.replace('\\', "/")),
```

Comment:
This is the expected behaviour, `/` must be used here as this is a repository-relative Git path.
Unfortunately there is no stronger type for it, maybe an oversight.